File: leetcode/java/74.search-a-2d-matrix.java
class Solution {
    // faltten 2 dem array
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        if (m == 0) return false;
        int n = matrix[0].length;
        if (n == 0) return false;

        int l = 0;
        int r = m * n;

        while (l < r) {
            int mid = (l + r) / 2;
            int _m = mid / n, _n = mid % n;
            if (matrix[_m][_n] == target) {
                return true;
            }
            if (matrix[_m][_n] > target) {
                r = mid;
            }
            if (matrix[_m][_n] < target) {
                l = mid + 1;
            }
        }
        return false;

    }
}

File: leetcode/js/74.search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
    let m = matrix.length;
    if (m == 0) return false;
    let n = matrix[0].length;
    if (n == 0) return false;

    // m line number
    // n number of elements every line has
    let i = 0;
    let j = m * n;
    while (i < j) {
        let mid = Math.floor((i + j) / 2);
        let value = matrix[Math.floor(mid / n)][mid % n];
        if (value == target) {
            return true;
        }

        if (value > target) {
            j = mid;
        }
        else {
            i = mid + 1;
        }
    }
    return false;
};
